repository-inmemory/src/main/java/tech/ydb/yoj/repository/test/inmemory/InMemoryDataShard.java

```diff
@@ -142,16 +143,39 @@ public synchronized void insert(long txId, long version, T entity) {
             throw new EntityAlreadyExistsException("Entity " + entity.getId() + " already exists");
         }
 
-        save(txId, entity);
+        save(txId, version, entity);
     }
 
-    public synchronized void save(long txId, T entity) {
+    public synchronized void save(long txId, long version, T entity) {
         InMemoryEntityLine entityLine = entityLines.computeIfAbsent(entity.getId(), __ -> new InMemoryEntityLine());
 
+        validateUniqueness(txId, version, entity);
         uncommited.computeIfAbsent(txId, __ -> new HashSet<>()).add(entity.getId());
+
         entityLine.put(txId, Columns.fromEntity(schema, entity));
     }
 
+    private void validateUniqueness(long txId, long version, T entity) {
+        List<Schema.Index> indexes = schema.getGlobalIndexes().stream()
+                .filter(Schema.Index::isUnique)
+                .toList();
+        for (Schema.Index index : indexes) {
+            Map<String, Object> entityIndexValues = buildIndexValues(index, entity);
+            for (InMemoryEntityLine line : entityLines.values()) {
+                Columns columns = line.get(txId, version);
+                if (columns != null && entityIndexValues.equals(buildIndexValues(index, columns.toSchema(schema)))) {
+                    throw new EntityAlreadyExistsException("Entity " + entity.getId() + " already exists");
+                }
+            }
+        }
+    }
+
+    private Map<String, Object> buildIndexValues(Schema.Index index, T entity) {
+        Map<String, Object> cells = schema.flatten(entity);
```

Comment:
Strictly speaking, `Schema.flatten()` is not guaranteed to return a mutable map (but in practice, it does). 

I would probably use `.entrySet().stream().filter(...).collect(toMap(...));` instead, to filter out fields not belonging to the index (and it would create a new map).

Reply:
Maybe `new HashMap<>(...);` =)? Same result - less code.

Reply:
Let's make a `new HashMap<>()` 👍